### ISSUE TITLE
unnecessary computed

### ISSUE THREAD

**Issue by chiaweilee** [vuejs/vue#8970]
### Version
2.5.17

### Reproduction link
[https://codesandbox.io/s/0mn6735xw0](https://codesandbox.io/s/0mn6735xw0)

### Steps to reproduce
When a Computed property is based on dependency which is only a property of an Object.

It just like this:

- An Array, a = [1, 2 ,3]
- Computed test property return a[0]
- Every time, this Array changed(push/unshift/pop...deep watch), computed test will run even this value of test never change.

Please see codesandbox demo (https://codesandbox.io/s/0mn6735xw0) for more infomation.

### What is expected?
this computed should be cached

### What is actually happening?
unnecessary computed

---
Normally, it's not a big problem.

It's huge when you try to compute and render large data like stock data which is an Object.

I did some job to try to resolve this problem.
https://github.com/chiaweilee/why-did-you-computed

**Comment by chiaweilee** [comment]
Well, I believe it's also a problem of project writer. Maybe we should not write code like this, we have better way to use property.

**Comment by chiaweilee** [comment]
> See #5776, this should be resolved by the new reactivity system in 3.0

Helpful! Thanks.

Indeed, It is not easy to resolved.

---

I am tring to do this:

transform the computed property into a FP like function, which work like this:

```
$computedTest () {
  return $transform(this.dependencyA, this.dependencyB, ...)(function (A, B) {
    ... // compute here
    return $result
  })
}
```

- This transform base on FP which use same arguments got same result
- Result cached(use tag of function and arguments)
- Return cached if exist(avoid compute)

It works for my project which did a lot of computes of stock data.

**Comment by posva** [comment]
Closing as this cannot be implemented with Object.defineProperty
